### DISCUSSION TITLE
File cannot be viewed, only downloaded (classed a raw)

### DISCUSSION THREAD

**Discussion by hearmonster** [community/community#22076] [New to GitHub]
<p>Similar to the thread “GitHub doesn’t show my code [view raw]”, my code files cannot be viewed directly and can only be downloaded (they are classed as ‘view raw’)</p>
<p>For instance, <a href="https://github.com/hearmonster/MQL4TradingSystem/blob/bc435e64eee7c577877f09f3110a0f41825c8111/MQL4/Include/PHFilePrinter.mqh" rel="nofollow noopener">here’s</a> one of my files</p>
<p>The file extension is an MQL4 .mqh (very similar to C)<br>
But other Repositories (e.g. <a href="https://github.com/rosasurfer/mt4-mql/blob/17b434f7a96cec650adfd7a93582a44172cba728/mql4/include/api.mqh" rel="nofollow noopener">this one</a>) have no trouble displaying these types of file directly.</p>
<p>Is there some sort of Repository (or individual file) configuration necessary to get these files recognised?  I’ve searched the community, googled but can’t find anything - except this thread I’ve mentioned above.</p>
<p>Thanks</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>You’ll notice that the repository that you linked to has a <a href="https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes"><code>.gitattributes</code> file</a>. This file is used to inform git that its defaults might not recognize a file type properly and to override them. I notice that in that <code>.gitattributes</code> file, there is <a href="https://github.com/rosasurfer/mt4-mql/blob/17b434f7a96cec650adfd7a93582a44172cba728/.gitattributes#L19">a line to force the file type for <code>.mqh</code> to <code>text</code></a> (as opposed to binary), as well as <a href="https://github.com/rosasurfer/mt4-mql/blob/17b434f7a96cec650adfd7a93582a44172cba728/.gitattributes#L33">force the character set to CP1252 for all files in the directory you linked to</a>. (git doesn’t do well when you’re using a character set other than UTF-8, which might be why those files are being recognized as binary instead of text.)</p>
<p>If you add a <code>.gitattributes</code> file to your repository and instruct git on how to properly interpret the contents of your files, then you should be able to make things appear the way you want.</p>
<p>I hope that helps!</p>

**Comment by hearmonster** [comment]
<p><strong>Thank you, Thank you!</strong><br>
Yeah, you were right. Not only did the .gitattributes supply the configuration I was looking for (i.e. how to handle my file extensions) but another <strong>thank you</strong> for bringing my attention to the file encodings.  (A couple of my files are indeed UTF-16! Tested with a new ANSI encoded file and viewing worked exactly as I expected.  No matter, I’ll copy the contents over to a new file with simple, ANSI encoding)</p>
<p>=o)</p>

**Comment by deepi110304** [comment]
can u say how does a raw file gets fixed
